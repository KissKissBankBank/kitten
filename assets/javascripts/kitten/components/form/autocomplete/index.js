import React, { useState, useRef, useEffect } from 'react'
import styled from 'styled-components'
import isFunction from 'lodash/fp/isFunction'
import PropTypes from 'prop-types'
import { pxToRem, stepToRem } from '../../../helpers/utils/typography'
import TYPOGRAPHY from '../../../constants/typography-config'
import COLORS from '../../../constants/colors-config'
import { VisuallyHidden } from '../../accessibility/visually-hidden'
import slugify from 'slugify'
import { Loader } from '../../graphics/animations/loader'
import classNames from 'classnames'

const itemHeight = 38
const maxItemsVisibled = 3
const borderSize = 'var(--border-width)'

const Wrapper = styled.div`
  display: flex;
  position: relative;

  .k-Form-Autocomplete__input {
    display: block;
    width: 100%;
    height: ${pxToRem(50)};
    box-sizing: border-box;
    border-radius: var(--border-radius-s);

    background: ${COLORS.background1};
    border: ${borderSize} solid var(--color-grey-400);
    padding: 0 ${pxToRem(15)};

    ${TYPOGRAPHY.fontStyles.light};
    font-size: ${stepToRem(-1)};
    line-height: 1.3;
    color: ${COLORS.font1};

    transition: border-color 0.4s;

    ::placeholder {
      color: ${COLORS.font2};
    }

    ::-moz-placeholder {
      color: ${COLORS.font2};
    }

    &:focus {
      border-color: var(--color-grey-500);
      outline: var(--outline-input);
      outline-offset: var(--outline-offset-input);
    }
  }
  &.k-Form-Autocomplete--error .k-Form-Autocomplete__input {
    border: var(--border-danger);
    color: ${COLORS.error3};

    :focus {
      border-color: var(--color-grey-500);
      color: ${COLORS.font1};
    }
  }

  .k-Form-Autocomplete__loader {
    display: flex;
    position: absolute;
    align-self: center;
    padding: 0 ${pxToRem(18)};
    z-index: 1;
    right: 0;
  }

  .k-Form-Autocomplete__icon {
    display: flex;
    position: absolute;
    align-self: center;
    padding: 0 ${pxToRem(18)};
    z-index: 1;
  }

  .k-Form-Autocomplete__suggestions {
    position: absolute;
    top: ${pxToRem(60)};
    left: 0;
    right: 0;
    overflow-y: auto;
    margin: 0;
    padding: 0;

    background: ${COLORS.background1};
    border: ${borderSize} solid var(--color-grey-400);
    border-radius: var(--border-radius-s);

    list-style: none;

    height: calc(${pxToRem(itemHeight)} * var(--Autocomplete-suggestions, 1));
    max-height: ${pxToRem(itemHeight * maxItemsVisibled)};
  }

  .k-Form-Autocomplete__suggestion__item {
    padding: ${pxToRem(10)} ${pxToRem(15)};

    ${TYPOGRAPHY.fontStyles.light};
    font-size: ${stepToRem(-1)};

    line-height: 1.3;
    color: ${COLORS.font1};

    &.k-Form-Autocomplete__suggestion__item--noresult {
      font-style: italic;
    }
    &:not(.k-Form-Autocomplete__suggestion__item--noresult) {
      cursor: pointer;
      transition: background-color 0.2s;

      :hover,
      :focus,
      :active {
        background-color: ${COLORS.background3};
      }

      &[aria-selected='true'] {
        background-color: ${COLORS.line1};
      }
    }
  }

  /* STATES */

  &.k-Form-Autocomplete--hasIcon-left {
    .k-Form-Autocomplete__input {
      padding-left: ${pxToRem(45)};
    }
    .k-Form-Autocomplete__icon {
      left: 0;
    }
  }

  &.k-Form-Autocomplete--hasIcon-right {
    .k-Form-Autocomplete__input {
      padding-right: ${pxToRem(45)};
    }
    .k-Form-Autocomplete__loader {
      padding-right: ${pxToRem(45)};
    }
    .k-Form-Autocomplete__icon {
      right: 0;
    }
  }

  &.k-Form-Autocomplete--disabled {
    .k-Form-Autocomplete__input {
      border: var(--color-grey-300);
      background-color: var(--color-grey-200);
      color: var(--color-grey-600);
    }

    .k-Form-Autocomplete__icon {
      & > svg [stroke]:not([stroke='none']) {
        stroke: ${COLORS.font2};
      }
      & > svg [fill]:not([fill='none']) {
        fill: ${COLORS.font2};
      }
    }
  }
`

export const Autocomplete = ({
  className,
  items: defaultItems,
  error,
  onChange,
  onBlur,
  onKeyDown,
  onSelect,
  icon,
  iconPosition,
  updateSuggestionsStrategy,
  isLoading,
  noResultMessage,
  shouldShowNoResultMessage,
  ...props
}) => {
  const [items, setItems] = useState(defaultItems)
  const [value, setValue] = useState(props.defaultValue)
  const [selectedItemIndex, setSelectedItemIndex] = useState(-1)
  const [showSuggestions, setShowSuggestions] = useState(false)
  const inputEl = useRef(null)
  const suggestionsEl = useRef(null)
  const showNoResultMessage = isFunction(shouldShowNoResultMessage)
    ? shouldShowNoResultMessage({ items, value })
    : shouldShowNoResultMessage

  useEffect(() => {
    updateSuggestions()
    setShowSuggestions(!!value)
  }, [value, defaultItems])

  const handleChange = e => {
    setValue(e.target.value)

    onChange(e)
  }

  const handleBlur = e => {
    setTimeout(() => {
      setShowSuggestions(false)
    }, 100)

    onBlur(e)
  }

  const handleKeyDown = e => {
    const enterKeyCode = 13
    const escKeyCode = 27
    const arrowUpKeyCode = 38
    const arrowDownKeyCode = 40

    if (e.keyCode === escKeyCode) setShowSuggestions(false)

    if (showSuggestions) {
      if (e.keyCode === arrowUpKeyCode) {
        e.preventDefault()
        prevSelectedItem()
      }

      if (e.keyCode === arrowDownKeyCode) {
        e.preventDefault()
        nextSelectedItem()
      }

      if (e.keyCode === enterKeyCode) {
        e.preventDefault()

        const selectedValue = items[selectedItemIndex]
        handleClickItem(selectedValue)()
      }
    }

    onKeyDown(e)
  }

  const handleClickItem = value => () => {
    if (!value) return

    inputEl.current.value = value
    inputEl.current.focus()

    setValue(value)
    setShowSuggestions(false)
    onSelect(value)
  }

  const updateSuggestions = () => {
    const search = `${value}`.toLowerCase()

    const newItems = updateSuggestionsStrategy
      ? updateSuggestionsStrategy({ items: defaultItems, value })
      : defaultItems.filter(
          item => item.toLowerCase().includes(search) && item !== value,
        )

    setItems(newItems)
    resetSelectedItem()
  }

  const prevSelectedItem = () => {
    const newIndex = selectedItemIndex - 1

    suggestionsScrollTop(newIndex)
    setSelectedItemIndex(newIndex < 0 ? 0 : newIndex)

    const scrollTop = newIndex * itemHeight
    suggestionsScrollTop(scrollTop)
  }

  const nextSelectedItem = () => {
    const newIndex = selectedItemIndex + 1
    const itemsLength = items.length - 1

    suggestionsScrollTop(newIndex)
    setSelectedItemIndex(newIndex >= itemsLength ? itemsLength : newIndex)

    const scrollTop = (newIndex - maxItemsVisibled + 1) * itemHeight
    suggestionsScrollTop(scrollTop)
  }

  const resetSelectedItem = () => setSelectedItemIndex(-1)

  const suggestionsScrollTop = value => {
    suggestionsEl.current.scrollTop = value > 0 ? value : 0
  }

  return (
    <Wrapper
      className={classNames('k-Form-Autocomplete', className, {
        [`k-Form-Autocomplete--hasIcon-${iconPosition}`]: !!icon,
        'k-Form-Autocomplete--disabled': props.disabled,
        'k-Form-Autocomplete--error': error,
      })}
    >
      <input
        {...props}
        ref={inputEl}
        type="text"
        role="combobox"
        autoComplete="off"
        onChange={handleChange}
        onBlur={handleBlur}
        onKeyDown={handleKeyDown}
        aria-owns={`${props.name}-results`}
        aria-expanded={showSuggestions ? items.length > 0 : null}
        aria-autocomplete="both"
        aria-activedescendant={
          items[selectedItemIndex]
            ? slugify(`${items[selectedItemIndex]}-${selectedItemIndex}`)
            : ''
        }
        className="k-Form-Autocomplete__input"
      />
      {isLoading && (
        <>
          <Loader
            className="k-Form-Autocomplete__loader"
            color={COLORS.font2}
          />
          <VisuallyHidden lang="en">loading</VisuallyHidden>
        </>
      )}
      {icon && (
        <span className="k-Form-Autocomplete__icon" aria-hidden="true">
          {React.cloneElement(icon, { width: 15, height: 15 })}
        </span>
      )}

      {showSuggestions &&
        items.length === 0 &&
        noResultMessage &&
        showNoResultMessage && (
          <ul
            ref={suggestionsEl}
            id={`${props.name}-results`}
            role="listbox"
            tabIndex="-1"
            style={{ '--Autocomplete-suggestions': '1' }}
            className="k-Form-Autocomplete__suggestions"
          >
            <li
              className="k-Form-Autocomplete__suggestion__item k-Form-Autocomplete__suggestion__item--noresult"
              role="option"
              tabIndex="-1"
            >
              {noResultMessage}
            </li>
          </ul>
        )}

      {showSuggestions && items.length > 0 && (
        <>
          <ul
            ref={suggestionsEl}
            id={`${props.name}-results`}
            role="listbox"
            tabIndex="-1"
            style={{ '--Autocomplete-suggestions': items.length }}
            className="k-Form-Autocomplete__suggestions"
          >
            {items.map((item, index) => (
              <li
                key={item + index}
                id={slugify(`${item}-${index}`)}
                onClick={handleClickItem(item)}
                role="option"
                aria-selected={selectedItemIndex === index}
                tabIndex="-1"
                className="k-Form-Autocomplete__suggestion__item"
              >
                {item}
              </li>
            ))}
          </ul>

          <VisuallyHidden lang="en" aria-live="assertive">
            {items.length} results are available.
          </VisuallyHidden>
        </>
      )}
    </Wrapper>
  )
}

Autocomplete.propTypes = {
  name: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  error: PropTypes.bool,
  icon: PropTypes.object,
  iconPosition: PropTypes.oneOf(['left', 'right']),
  updateSuggestionsStrategy: PropTypes.func,
  noResultMessage: PropTypes.string,
  shouldShowNoResultMessage: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.func,
  ]),
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  onKeyDown: PropTypes.func,
  onSelect: PropTypes.func,
  isLoading: PropTypes.bool,
}

Autocomplete.defaultProps = {
  error: false,
  shouldShowNoResultMessage: true,
  iconPosition: 'left',
  onChange: () => {},
  onBlur: () => {},
  onKeyDown: () => {},
  onSelect: () => {},
  isLoading: false,
}
