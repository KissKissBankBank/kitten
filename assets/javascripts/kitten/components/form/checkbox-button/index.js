import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import styled from 'styled-components'
import { Text } from '../../typography/text'
import { pxToRem, stepToRem } from '../../../helpers/utils/typography'
import { mq } from '../../../constants/screen-config'

const StyledCheckboxButton = styled.div`
  font-size: ${stepToRem(-1)};
  line-height: ${pxToRem(18)};
  margin: ${pxToRem(10)} 0;

  &.k-Form-CheckboxButton--disabled input,
  &.k-Form-CheckboxButton--disabled label {
    cursor: not-allowed;
  }

  .k-Form-CheckboxButton__label {
    display: flex;
    align-items: center;
    align-self: flex-start;
    flex-wrap: wrap;
    cursor: pointer;
    gap: ${pxToRem(5 + 2)} ${pxToRem(15)};
    box-sizing: border-box;
    border-radius: var(--border-radius-s);
    border: var(--border);
    transition: border var(--transition), background var(--transition);
  }

  .k-Form-CheckboxButton__label::before {
    display: block;

    content: '';
    box-sizing: border-box;
    display: inline-block;
    flex-shrink: 0;
    width: ${pxToRem(20)};
    height: ${pxToRem(20)};
    border-radius: var(--border-radius-s);

    background-color: var(--color-grey-000);
    border: var(--border);
    cursor: pointer;
    transition: border-color var(--transition),
      background-color var(--transition);
  }

  .k-Form-CheckboxButton__input:focus 
    & + .k-Form-CheckboxButton__label {
      border: var(--border-active-primary);
    }
  }

  .k-Form-CheckboxButton__input:checked + .k-Form-CheckboxButton__label::before,
  .k-Form-CheckboxButton__input:indeterminate
    + .k-Form-CheckboxButton__label::before {
    background-color: var(--color-primary-500);
    background-repeat: no-repeat;
    background-position: center;
    border: var(--border-active-primary);
  }

  .k-Form-CheckboxButton__input:checked
    + .k-Form-CheckboxButton__label::before {
    background-image: url("data:image/svg+xml,%3Csvg width='11' height='8' viewBox='0 0 11 8' xmlns='http://www.w3.org/2000/svg' fill='%23fff'%0A%3E%3Cpath d='M4.44 7.596L3.024 6.182 8.682.525l1.414 1.414z' /%3E%3Cpath d='M4.44 7.596L.903 4.06l1.414-1.414 3.536 3.536z' /%3E%3C/svg%3E%0A");
  }

  .k-Form-CheckboxButton__input:indeterminate
    + .k-Form-CheckboxButton__label::before {
    background-image: url("data:image/svg+xml,%3Csvg width='6' height='2' viewBox='0 0 6 2' xmlns='http://www.w3.org/2000/svg' fill='%23fff'%0A%3E%3Crect width='6' height='2' /%3E%3C/svg%3E%0A");
  }

  .k-Form-CheckboxButton__input {
    appearance: none;
    position: absolute;
    opacity: 0;
    cursor: pointer;
  }

  .k-Form-CheckboxButton__input:hover:not(:disabled) {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-primary-300);
      ::before {
        border-color: var(--color-primary-300);
      }
    }
  }

  .k-Form-CheckboxButton__input:focus:not(:disabled) {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-primary-500);     
      ::before {
        border-color: var(--color-primary-500);
      }
    }
  }

  .k-Form-CheckboxButton__input:active:not(:disabled) {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-primary-500);
      ::before {
        border-color: var(--color-primary-500);
      }
    }
  }

  .k-Form-CheckboxButton__input:checked:not(:disabled) {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-primary-500);
      background-color: var(--color-primary-100);
    }
  }

  /* Disabled */

  .k-Form-CheckboxButton__input:disabled {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-grey-300);
      color: var(--color-grey-600);

      &::before {
        border-color: var(--color-grey-400);
        background-color: var(--color-grey-000);
      }
    }
  }

  .k-Form-CheckboxButton__input:checked:disabled {
    & + .k-Form-CheckboxButton__label {
      border-color: var(--color-grey-600);
      color: var(--color-grey-600);
      background-color: var(--color-grey-200);

      &::before {
        border-color: var(--color-grey-500);
        background-color: var(--color-grey-500);
      }
    }
  }

  &.k-Form-CheckboxButton--error {
    .k-Form-CheckboxButton__input:not(:disabled) {
      & + .k-Form-CheckboxButton__label {
        background-color: var(--color-grey-000);
        border-color: var(--color-danger-300);

        &::before {
          border-color: var(--color-danger-500);
        }
      }

      &:checked + .k-Form-CheckboxButton__label::before {
        background-color: var(--color-danger-500);
      }

      &:checked:focus + .k-Form-CheckboxButton__label::before {
        box-shadow: 0 0 ${pxToRem(3)} var(--color-danger-500);
      }
    }
  }
  &.k-Form-CheckboxButton--small .k-Form-CheckboxButton__label {
    padding: ${pxToRem(10 - 1)} ${pxToRem(15)};
    min-height: ${pxToRem(40)};
  }

  &.k-Form-CheckboxButton--medium .k-Form-CheckboxButton__label {
    padding: ${pxToRem(20 - 1)} ${pxToRem(20)};
    min-height: ${pxToRem(60)};
  }

  &.k-Form-CheckboxButton--large .k-Form-CheckboxButton__label {
    padding: ${pxToRem(20 - 1)} ${pxToRem(20)};
    min-height: ${pxToRem(60)};
  }

  @media ${mq.tabletAndDesktop} {
    &.k-Form-CheckboxButton--large .k-Form-CheckboxButton__label {
      padding: ${pxToRem(30 - 1)} ${pxToRem(30)};
      min-height: ${pxToRem(80)};
    }
  }

  .k-Form-CheckboxButton__labelText {
    flex-basis: calc(100% - ${pxToRem(20 + 15)});
  }

  .k-Form-CheckboxButton__content {
    flex: 1 0 calc(100% - ${pxToRem(20 + 15)});
    margin: 0 0 0 ${pxToRem(20 + 15)};
  }
`

export const CheckboxButton = ({
  className,
  id,
  children,
  inputClassName,
  error,
  disabled,
  textProps,
  labelProps,
  onLabelClick,
  indeterminate,
  label,
  size,
  ...inputProps
}) => {
  const inputElement = useRef(null)

  useEffect(() => {
    if (inputElement.current != null) {
      inputElement.current.indeterminate = indeterminate
    }
  }, [indeterminate])

  return (
    <StyledCheckboxButton
      className={classNames(
        'k-Form-CheckboxButton',
        `k-Form-CheckboxButton--${size}`,
        {
          'k-Form-CheckboxButton--error': error,
          'k-Form-CheckboxButton--disabled': disabled,
        },
        className,
      )}
    >
      <input
        ref={inputElement}
        id={id}
        type="checkbox"
        className={classNames('k-Form-CheckboxButton__input', inputClassName)}
        disabled={disabled}
        {...inputProps}
      />

      {!label && children && (
        <Text
          tag="label"
          htmlFor={id}
          onClick={onLabelClick}
          weight="500"
          {...labelProps}
          {...textProps}
          className={classNames(
            'k-Form-CheckboxButton__label',
            textProps?.className,
            labelProps?.className,
          )}
        >
          <span className="k-Form-CheckboxButton__labelText">{children}</span>
        </Text>
      )}

      {label && !children && (
        <Text
          tag="label"
          htmlFor={id}
          onClick={onLabelClick}
          weight="500"
          {...labelProps}
          {...textProps}
          className={classNames(
            'k-Form-CheckboxButton__label',
            textProps?.className,
            labelProps?.className,
          )}
        >
          <span className="k-Form-CheckboxButton__labelText">{label}</span>
        </Text>
      )}

      {label && children && (
        <label
          className={classNames('k-Form-CheckboxButton__label')}
          htmlFor={id}
          onClick={onLabelClick}
        >
          <Text
            weight="500"
            {...labelProps}
            className={classNames(
              'k-Form-CheckboxButton__labelText',
              labelProps?.className,
            )}
          >
            {label}
          </Text>
          <Text
            {...textProps}
            className={classNames(
              'k-Form-CheckboxButton__content',
              textProps?.className,
            )}
          >
            {children}
          </Text>
        </label>
      )}
    </StyledCheckboxButton>
  )
}

CheckboxButton.defaultProps = {
  id: 'checkbox',
  textProps: {},
  labelProps: {},
  indeterminate: false,
  error: false,
  disabled: false,
  onLabelClick: () => {},
  size: 'medium',
}

CheckboxButton.propTypes = {
  textProps: PropTypes.object,
  labelProps: PropTypes.object,
  indeterminate: PropTypes.bool,
  error: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  children: PropTypes.node,
  label: PropTypes.string,
  onLabelClick: PropTypes.func,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
}
