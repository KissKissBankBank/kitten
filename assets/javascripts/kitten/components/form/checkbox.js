import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

export class Checkbox extends React.Component {
  render() {
    const { className,
            id,
            children,
            inputClassName,
            error,
            ...inputProps } = this.props

    const checkboxInputClassNames = classNames(
      'k-Checkbox__input',
      inputClassName,
      { 'is-error': error },
    )

    return (
      <div className={ classNames('k-Checkbox', className) }>
        <input id={ id }
               type="checkbox"
               className={ checkboxInputClassNames }
               { ...inputProps } />

        <label htmlFor={ id } className="k-Checkbox__label">
          { children }
        </label>
      </div>
    )
  }
}

Checkbox.defaultProps = {
  children: 'Filter 1',
}
