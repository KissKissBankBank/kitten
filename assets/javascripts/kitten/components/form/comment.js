import React, { Component, Fragment } from 'react'
import Radium, { StyleRoot } from 'radium'
import PropTypes from 'prop-types'
import { Marger as MargerBase } from 'kitten/components/layout/marger'
import { CommentAvatar } from 'kitten/components/form/comment-avatar'
import { Text } from 'kitten/components/typography/text'
import { ScreenConfig } from 'kitten/constants/screen-config'
import COLORS from 'kitten/constants/colors-config'

const Marger = Radium(MargerBase)

export class Comment extends Component {
  static PropTypes = {
    text: PropTypes.string.isRequired,
    ownerName: PropTypes.string.isRequired,
    avatarImgProps: PropTypes.object.isRequired,
    commentDate: PropTypes.string.isRequired,
  }

  render() {
    const { avatarProps, commentDate } = this.props

    return (
      <StyleRoot>
        <div style={styles.grid}>
          <CommentAvatar
            avatarImgProps={avatarProps}
            commentDate={commentDate}
          />
          {this.renderComment()}
        </div>
      </StyleRoot>
    )
  }

  renderComment() {
    const { text, ownerName } = this.props

    return (
      <div style={styles.comment}>
        <Marger bottom="1">
          <Text color="font1" size="tiny" weight="regular">
            {ownerName}
          </Text>
        </Marger>
        <Marger top="1">{text}</Marger>
        <span style={styles.comment.arrow} />
      </div>
    )
  }
}

const styles = {
  grid: {
    display: 'flex',
  },
  comment: {
    position: 'relative',
    borderWidth: '2px',
    backgroundColor: COLORS.background3,
    borderColor: COLORS.background3,
    color: COLORS.font1,
    padding: '30px',
    fontSize: '16px',
    marginLeft: 20,

    [`@media (min-width: ${ScreenConfig['S'].min}px)`]: {
      marginLeft: 35,
    },

    arrow: {
      position: 'absolute',
      top: '30px',
      display: 'block',
      width: 0,
      height: 0,
      borderWidth: '10px',
      borderStyle: 'solid',
      borderColor: 'transparent',
      borderRightColor: COLORS.background3,
      left: '-20px',
    },
  },
}
