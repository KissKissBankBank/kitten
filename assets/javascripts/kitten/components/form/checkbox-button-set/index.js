import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classNames from 'classnames'
import { CheckboxButton } from '../../form/checkbox-button'
import { pxToRem } from '../../../helpers/utils/typography'
import { Label } from '../../form/label'

const StyledCheckboxButtonSet = styled.fieldset`
  margin: 0;
  padding: 0;
  border: 0;
  display: flex;
  flex-direction: column;
  line-height: 1.3rem;

  & > legend {
    padding: 0;
  }

  .k-Form-CheckboxButtonSet__label {
    margin-bottom: ${pxToRem(10)};
  }

  .k-Form-CheckboxButtonSet__checkbox {
    margin: ${pxToRem(5)} 0;

    &:first-of-type {
      margin-top: 0;
    }
    &:last-of-type {
      margin-bottom: 0;
    }
  }
`

export const CheckboxButtonSet = ({
  items,
  disabled,
  className,
  name,
  error,
  label,
  children,
  ...props
}) => (
  <StyledCheckboxButtonSet
    className={classNames('k-Form-CheckboxButtonSet', className)}
    disabled={disabled}
    {...props}
  >
    {label && (
      <Label tag="legend" className="k-Form-CheckboxButtonSet__label">
        {label}
      </Label>
    )}
    {children && !label && <legend>{children}</legend>}
    {items.map(({ id, className, ...itemProps }) => (
      <CheckboxButton
        id={id}
        error={error}
        name={name}
        key={id}
        {...itemProps}
        className={classNames('k-Form-CheckboxButtonSet__checkbox', className)}
      />
    ))}
  </StyledCheckboxButtonSet>
)

CheckboxButtonSet.propTypes = {
  name: PropTypes.string.isRequired,
  error: PropTypes.bool,
  label: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      id: PropTypes.string.isRequired,
      defaultChecked: PropTypes.bool,
      children: PropTypes.node,
    }),
  ),
  disabled: PropTypes.bool,
}

CheckboxButtonSet.defaultProps = {
  name: 'checkboxSet',
  error: false,
  label: null,
  items: [
    {
      label: null,
      children: null,
      defaultChecked: true,
      id: 'myCheckboxButton', // Replace by a real value
    },
  ],
  disabled: false,
}
