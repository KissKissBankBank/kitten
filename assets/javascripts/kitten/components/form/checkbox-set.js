import React from 'react'
import classNames from 'classnames'
import { Checkbox } from 'kitten/components/form/checkbox'

export class CheckboxSet extends React.Component {
  constructor(props) {
    super(props)
    this.renderCheckbox = this.renderCheckbox.bind(this)
  }

  renderCheckbox(item, index) {
    const { className, ...itemProps } = item
    const checkboxClassName = classNames('k-CheckboxSet__checkbox', className)

    return (
      <Checkbox
        className={checkboxClassName}
        error={this.props.error}
        name={this.props.name}
        key={item.id}
        {...itemProps}
      />
    )
  }

  render() {
    const { items, className, name, error, ...checkboxSetProps } = this.props
    const checkboxSetClassName = classNames('k-CheckboxSet', className)

    return (
      <div className={checkboxSetClassName} {...checkboxSetProps}>
        {this.props.items.map(this.renderCheckbox)}
      </div>
    )
  }
}

CheckboxSet.defaultProps = {
  name: 'checkboxSet[]',
  error: false,
  items: [
    {
      label: 'filter 1',
      defaultChecked: true,
      id: 'myCheckbox', // Replace by a real value
    },
  ],
}
