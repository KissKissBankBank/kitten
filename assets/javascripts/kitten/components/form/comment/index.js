import React from 'react'
import { Comment as DefaultComment } from '../../../components/comments/comment'
import { Deprecated } from '../../../helpers/utils/deprecated'

export const Comment = props => {
  const message = `You can use the Comment instead.
    For more information: https://github.com/KissKissBankBank/../../blob/master/assets/javascripts/../../components/comments/comment.js`

  return (
    <Deprecated warningMessage={message}>
      <DefaultComment {...props} />
    </Deprecated>
  )
}
