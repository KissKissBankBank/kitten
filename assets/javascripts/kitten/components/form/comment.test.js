import React from 'react'
import renderer from 'react-test-renderer'
import { Comment } from 'kitten/components/form/comment'

const createMockMediaMatcher = matches => () => ({
  matches,
  addListener: () => {},
  removeListener: () => {},
})

describe('<Comment />', () => {
  let originalMatchMedia
  let component

  beforeEach(() => {
    originalMatchMedia = window.matchMedia
  })

  afterEach(() => {
    window.matchMedia = originalMatchMedia
  })

  it('should match its empty snapshot', () => {
    window.matchMedia = createMockMediaMatcher(false)

    const tree = renderer
      .create(
        <Comment
          avatarImgProps={{
            src: '#image',
            alt: 'Image alt',
          }}
          commentDate="Custom date"
          text="Custom text"
          ownerName="Custom name"
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should match a snapshot with a complex bottom notes', () => {
    window.matchMedia = createMockMediaMatcher(false)
    const BottomNotes = () => (
      <div>
        <a href="#" style={{ color: '#000' }}>
          Delete this comment
        </a>
        <a href="#" style={{ color: '#000' }}>
          Notify abuse
        </a>
      </div>
    )

    const tree = renderer
      .create(
        <Comment
          avatarImgProps={{
            src: '#image',
            alt: 'Image alt',
          }}
          commentDate="Custom date"
          text="Custom text"
          ownerName="Custom name"
          bottomNotes={<BottomNotes />}
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
