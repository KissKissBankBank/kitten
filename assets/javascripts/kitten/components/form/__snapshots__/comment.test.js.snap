// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> should match a snapshot with a complex bottom notes 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "marginBottom": "0.625rem",
          "marginTop": null,
        }
      }
    >
      <span
        className="k-ButtonImage k-ButtonImage--huge k-ButtonImage--withoutPointerEvents"
        tabIndex={null}
      >
        <img
          alt="Image alt"
          className="k-ButtonImage__img"
          src="#image"
        />
      </span>
    </div>
    <span
      className="k-u-color-font1 k-u-size-micro k-u-weight-light"
    >
      Custom date
    </span>
  </div>
  <div
    className="rmq-993edb07"
    data-radium={true}
    style={
      Object {
        "marginLeft": 20,
        "position": "relative",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#f6f6f6",
          "borderColor": "#f6f6f6",
          "borderWidth": 2,
          "color": "#222",
          "fontSize": 16,
          "padding": 30,
        }
      }
    >
      <div
        style={
          Object {
            "marginBottom": "0.625rem",
            "marginTop": null,
          }
        }
      >
        <span
          className="k-u-color-font1 k-u-size-tiny k-u-weight-regular"
        >
          Custom name
        </span>
      </div>
      <div
        style={
          Object {
            "marginBottom": null,
            "marginTop": "0.625rem",
          }
        }
      >
        <span
          className="k-u-color-font1 k-u-size-default k-u-weight-light"
        >
          Custom text
        </span>
      </div>
      <span
        className="rmq-e4f3f344"
        data-radium={true}
        style={
          Object {
            "borderColor": "transparent",
            "borderRightColor": "#f6f6f6",
            "borderStyle": "solid",
            "borderWidth": 10,
            "display": "block",
            "height": 0,
            "left": -20,
            "position": "absolute",
            "top": 20,
            "width": 0,
          }
        }
      />
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "@media (min-width: 640px){ .rmq-e4f3f344{top: 35px !important;}}",
          }
        }
      />
    </div>
    <p
      className="k-u-color-font1 k-u-size-micro k-u-weight-bold"
      data-radium={true}
      style={
        Object {
          "paddingLeft": 30,
        }
      }
    >
      <div>
        <a
          href="#"
          style={
            Object {
              "color": "#000",
            }
          }
        >
          Delete this comment
        </a>
        <a
          href="#"
          style={
            Object {
              "color": "#000",
            }
          }
        >
          Notify abuse
        </a>
      </div>
    </p>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": "@media (min-width: 640px){ .rmq-993edb07{margin-left: 35px !important;}}",
        }
      }
    />
  </div>
</div>
`;

exports[`<Comment /> should match its empty snapshot 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "marginBottom": "0.625rem",
          "marginTop": null,
        }
      }
    >
      <span
        className="k-ButtonImage k-ButtonImage--huge k-ButtonImage--withoutPointerEvents"
        tabIndex={null}
      >
        <img
          alt="Image alt"
          className="k-ButtonImage__img"
          src="#image"
        />
      </span>
    </div>
    <span
      className="k-u-color-font1 k-u-size-micro k-u-weight-light"
    >
      Custom date
    </span>
  </div>
  <div
    className="rmq-993edb07"
    data-radium={true}
    style={
      Object {
        "marginLeft": 20,
        "position": "relative",
      }
    }
  >
    <div
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#f6f6f6",
          "borderColor": "#f6f6f6",
          "borderWidth": 2,
          "color": "#222",
          "fontSize": 16,
          "padding": 30,
        }
      }
    >
      <div
        style={
          Object {
            "marginBottom": "0.625rem",
            "marginTop": null,
          }
        }
      >
        <span
          className="k-u-color-font1 k-u-size-tiny k-u-weight-regular"
        >
          Custom name
        </span>
      </div>
      <div
        style={
          Object {
            "marginBottom": null,
            "marginTop": "0.625rem",
          }
        }
      >
        <span
          className="k-u-color-font1 k-u-size-default k-u-weight-light"
        >
          Custom text
        </span>
      </div>
      <span
        className="rmq-e4f3f344"
        data-radium={true}
        style={
          Object {
            "borderColor": "transparent",
            "borderRightColor": "#f6f6f6",
            "borderStyle": "solid",
            "borderWidth": 10,
            "display": "block",
            "height": 0,
            "left": -20,
            "position": "absolute",
            "top": 20,
            "width": 0,
          }
        }
      />
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": "@media (min-width: 640px){ .rmq-e4f3f344{top: 35px !important;}}",
          }
        }
      />
    </div>
    
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": "@media (min-width: 640px){ .rmq-993edb07{margin-left: 35px !important;}}",
        }
      }
    />
  </div>
</div>
`;
